### ISSUE TITLE
Crash when moving PartialComponent to Component

### ISSUE THREAD

**Issue by krojew** [google/fruit#67]
EDIT: The description was revised due to further investigation.

fruit requires the whole ParialComponent chain to be alive until the creation of an actual component. Due to PartialComponent being movable, yet keeping a reference to component store from a previous PartialComponent, we can easily achieve a scenario when code passing around partial components by value builds without any warning and crashes at runtime.

**Comment by krojew** [comment]
After some digging it looks to be related to moving - PartialComponent is movable and moves its storage, which then simply keeps the reference to previous storage which is about to expire, thus causing undefined 
behavior.

**Comment by poletti-marco** [comment]
Hi, yes, this is expected (but I understand that it might be surprising).
You should not create variables with type PartialComponent.

See the comment on PartialComponent here:
https://github.com/google/fruit/blob/master/include/fruit/component.h


"Client code should never write `PartialComponent'; always start the
construction of a component with
fruit::createComponent(), and end it by casting the PartialComponent to the
desired Component (often done implicitly
by returning a PartialComponent from a function that has Component<...> as
the return type).
[...]
Note that no variable of type PartialComponent has been declared; this
class should only be used for temporary
values."

**Comment by krojew** [comment]
This should be enforced at compile time, by making PartialComponent non-movable.

**Comment by poletti-marco** [comment]
I wish we could enforce this at compile-time, but removing the Component constructor that takes a PartialComponent&& would also make the other example (where we immediately convert to Component) also fail to compile AFAICT.

I don't see a way to fail compilation in one case but not the other.
Any ideas are welcome though.

**Comment by poletti-marco** [comment]
Oh, and deleting the move constructor of PartialComponent would not prevent either of those AFAICT.

**Comment by krojew** [comment]
Actually deleting the move ctor of PartialComponent would solve this case, because there would be no way to move it implicitly or explicitly. At the moment you can, for example, pass it by value resulting in an undefined behavior. Unfortunately, we stumbled upon this very issue and it took two days of debugging random crashes to finally pinpoint the issue.

From a logical point of view, PartialComponent should not be movable anyway, as it stands right now. After all, nothing is being moved - the only effect is creating a (hidden) reference to something which is about to be destroyed. The natural assumption is to either move the actual contents, including ownership of data, or to make a copy, hence not having any fear of accessing something which is about to expire. PartialComponent moves don't result in any of these, therefore being very misleading.

**Comment by poletti-marco** [comment]
> Actually deleting the move ctor of PartialComponent would solve this case, because there would be no way to move it implicitly or explicitly. 

I don't think it would. Note that the move constructor is *not* invoked when using `std::move()`, unless you're constructing a 2nd instance of PartialComponent with the result of `std::move()`. `std::move()` by itself is just a cast to a rvalue reference, and it's always allowed. Let me explain with an example:


```
#include <utility>

class X {
public:
    X() = default;

    X(const X&) = delete;
    X(X&&) = delete;
    X& operator=(const X&) = delete;
    X& operator=(X&&) = delete;
};

class Y {
public:
    Y(X&&) {}
};

int main() {
    X x;
    Y y = std::move(x);
    (void) y;
    return 0;
}
```

Even though X is non-movable and the code calls `std::move()`, the above compiles cleanly.
Only the following would not compile:

    X x2 = std::move(x);

**Comment by krojew** [comment]
The issue is not with moving PartialComponent to Component. It's with moving between PartialComponents - your second case.

**Comment by krojew** [comment]
Maybe let me provide an example:

```
auto pc = fruit::createComponent();
auto foo = [](auto pc) {
    return pc;
}

fruit::Component<...> c = foo(c); // undefined behavior here
```

**Comment by poletti-marco** [comment]
You're right, in the case where `PartialComponent` is returned by a function, making it non-movable would help catch this. I've now submitted that change as commit ed5ec821287f0b4cf0fe784f0b98e055a909bf1d and I'll create a bugfix release of Fruit with this fix shortly.

Thanks for your report and for persisting in your explanation.

**Comment by poletti-marco** [comment]
Now released v3.2.0 with this: https://github.com/google/fruit/releases/tag/v3.2.0
The binary packages are being built and will be available shortly (most of them already are).
